Refactor cross compile assistance

runesoerensen opened the pull request:
We currently support targeting `x86_64-unknown-linux-musl` and `aarch64-unknown-linux-musl` on macOS and Linux, but the cross compile assistance provided is inconsistent depending on the host's CPU architecture. This refactoring aims to improve that across all supported `target_triple` and host OS/architecture combinations by providing helpful assistance for supported scenarios.

* Fixed https://github.com/heroku/libcnb.rs/issues/728 by improving the help text to include accurate advice for the `target_triple` relative to the current host OS *and* architecture.
* Fixed https://github.com/heroku/libcnb.rs/issues/726 by adding support for targeting `x86_64-unknown-linux-musl` from AARCH64 Linux.
* Improves maintainability by sharing CrossCompileAssistance::Configuration initialization with relevant environment variables (when appropriate) in one place (which will be useful if we want to set additional environment variables in the future, i.e. https://github.com/heroku/libcnb.rs/issues/725).
* Fixes part of https://github.com/heroku/libcnb.rs/issues/724 by adding support for targeting `aarch64-unknown-linux-musl` from AARCH64 Linux. The gcc binary used when targeting the same architecture is now `musl-gcc` on both `aarch64` and `x86-64`.
    * However, the binary used when the architectures don't match is still the `*-gnu-gcc` variant (as for instance `aarch64-linux-musl-gcc` is only available on aarch64 Linux (this requires further investigation, and a separate issue should be created for that)).
* Adds support for targeting `aarch64-unknown-linux-musl` from aarch64 macOS.
* Adds strict host architecture checks to ensure that `CrossCompileAssistance::Configuration` or `CrossCompileAssistance::HelpText` is only returned when supported. Otherwise `CrossCompileAssistance::NoAssistance` is returned.

It also:
* Drops support for a gcc binary (`x86_64-linux-musl-gcc`) installed by a previously recommended macOS toolchain. Since the current toolchain has been recommended for nearly 2 years it's probably safe to assume (or fair to require) that users have update to the recommended toolchain. 
The brew formula/revision has also been bumped, so the missing dependencies mentioned here https://github.com/heroku/libcnb.rs/pull/312#issuecomment-1038044636 are now included

Malax commented on libcnb-package/src/cross_compile.rs:
```diff
@@ -111,5 +88,12 @@ pub enum CrossCompileAssistance {
 const AARCH64_UNKNOWN_LINUX_MUSL: &str = "aarch64-unknown-linux-musl";
 const X86_64_UNKNOWN_LINUX_MUSL: &str = "x86_64-unknown-linux-musl";
+
+// Constants for OS and ARCH
```
Nit:
```suggestion
// Constants for `std::env::consts::OS` and `std::env::consts::ARCH`
```

Malax commented on libcnb-package/src/cross_compile.rs:
```diff
@@ -7,95 +8,71 @@ use which::which;
 
-You will also need to install the Rust target, using:
-rustup target add aarch64-unknown-linux-musl",
-            )),
-        }
-    } else {
-        CrossCompileAssistance::NoAssistance
+{help_text}
+            
+You will also need to install the Rust target:
+rustup target add {target}",
+            consts::ARCH,
+            consts::OS
+        )),
```
This seems like a good use-case for [indoc](https://docs.rs/indoc/latest/indoc/index.html). We use that crate a lot and it helps making these multi-line strings much nicer in the code as it will strip away the common whitespace for your. And since it's a macro, there is no impact on the resulting code.

Malax commented on libcnb-package/src/cross_compile.rs:
```diff
@@ -7,95 +8,71 @@ use which::which;
-        // an older version cross_compile_assistance which suggested installing a different
-        // toolchain.
-        let possible_gcc_binary_names = ["x86_64-unknown-linux-musl-gcc", "x86_64-linux-musl-gcc"];
+    let target = target_triple.as_ref();
```
Rust allows to re-declare variables with the same name. This would allow us to do: 

```suggestion
    let target_triple = target_triple.as_ref();
```

No need to call [`AsRef::as_ref`](https://doc.rust-lang.org/std/convert/trait.AsRef.html#tymethod.as_ref) everywhere but also no loss of information calling the target triple just `target` which can be ambiguous to readers.

Malax commented on libcnb-package/src/cross_compile.rs:
```diff
@@ -7,95 +8,71 @@ use which::which;
-        let possible_gcc_binary_names = ["x86_64-unknown-linux-musl-gcc", "x86_64-linux-musl-gcc"];
+    let target = target_triple.as_ref();
+    let (gcc_path, help_text) = match (target, consts::OS, consts::ARCH) {
+        (AARCH64_UNKNOWN_LINUX_MUSL, OS_LINUX, ARCH_X86_64) => (
+            "aarch64-linux-gnu-gcc",
+            "To install an aarch64 cross-compiler on Ubuntu:\nsudo apt-get install g++-aarch64-linux-gnu",
+        ),
+        (AARCH64_UNKNOWN_LINUX_MUSL, OS_MACOS, _) => (
+            "aarch64-unknown-linux-musl-gcc",
+            "To install an aarch64 cross-compiler on macOS:\nbrew install messense/macos-cross-toolchains/aarch64-unknown-linux-musl",
+        ),
+        (AARCH64_UNKNOWN_LINUX_MUSL, OS_LINUX, ARCH_AARCH64) | (X86_64_UNKNOWN_LINUX_MUSL, OS_LINUX, ARCH_X86_64) => (
+            "musl-gcc",
+            "To install musl-tools on Ubuntu:\nsudo apt-get install musl-tools",
+        ),
+        (X86_64_UNKNOWN_LINUX_MUSL, OS_LINUX, ARCH_AARCH64) => (
+            "x86_64-linux-gnu-gcc",
+            "To install an x86_64 cross-compiler on Ubuntu:\nsudo apt-get install g++-x86_64-linux-gnu",
+        ),
+        (X86_64_UNKNOWN_LINUX_MUSL, OS_MACOS, _) => (
+            "x86_64-unknown-linux-musl-gcc",
+            "To install an x86_64 cross-compiler on macOS:\nbrew install messense/macos-cross-toolchains/x86_64-unknown-linux-musl",
+        ),
```
For these multiline help texts, using [indoc](https://docs.rs/indoc/latest/indoc/index.html) and writing them with actual newlines in the code should improve readability a little bit.

Malax commented on libcnb-package/src/cross_compile.rs:
```diff
@@ -7,95 +8,71 @@ use which::which;
-            )), |gcc_binary_path| {
+    match which(gcc_path) {
+        Ok(_) => {
+            if gcc_path == "musl-gcc" {
+                CrossCompileAssistance::Configuration {
+                    cargo_env: Vec::new(),
+                }
```
Do we actually need this branch or could the code in the `else` branch work for `musl-gcc` as well? If we do need this branch, can we add a short comment why?

Malax commented on libcnb-package/src/cross_compile.rs:
```diff
@@ -7,95 +8,71 @@ use which::which;
+            "aarch64-linux-gnu-gcc",
+            "To install an aarch64 cross-compiler on Ubuntu:\nsudo apt-get install g++-aarch64-linux-gnu",
+        ),
+        (AARCH64_UNKNOWN_LINUX_MUSL, OS_MACOS, _) => (
```
Since the `_` will match any value, wouldn't it be better to put `ARCH_X86_64 | ARCH_AARCH64` there to be explicit and to ensure we'll end up in the catch-all for unknown architectures?

(There is another `_` match later as well)

runesoerensen replied:
For sure -- addressed in https://github.com/heroku/libcnb.rs/pull/769/commits/cab120b23e7965ef932119fbe813df3ce74e31f5

Malax reviewed (changes requested):
Thanks for the PR! 🎉 

I gave this a skim-read to provide some early feedback since this is one of your first PRs around here. I understand that this is a draft still.

Apart from the mostly minor code comments, I was missing some context due to the fact there is not PR description and the title being very vague (what does it fix, is there an existing issue somewhere?) Again, this is a draft PR and I write my descriptions at the very end as well. If that was to come later - ignore my comment! :)

There also seem to be some slight different in packages recommended. Have you verified those on the respective systems? For example, `libc6-dev-arm64-cross` is no longer mentioned. It's possible it is not needed, but from the information I have it could also be an oversight. :)

runesoerensen commented:
> Apart from the mostly minor code comments, I was missing some context due to the fact there is not PR description and the title being very vague (what does it fix, is there an existing issue somewhere?) Again, this is a draft PR and I write my descriptions at the very end as well. If that was to come later - ignore my comment! :)

I have added a more elaborate description now with links to relevant issues. Merging this pull request should close those issues, but let me know if it's better to unlink them and manually close the issues with reference to this pull request!

> There also seem to be some slight different in packages recommended. Have you verified those on the respective systems? For example, `libc6-dev-arm64-cross` is no longer mentioned. It's possible it is not needed, but from the information I have it could also be an oversight. :)

I initially removed the `libc6-dev-arm64-cross` package from the help text as during testing it was installed along with `g++-aarch64-linux-gnu` - but only as a recommended package - https://github.com/heroku/libcnb.rs/pull/769/commits/c5a046b0d7699e95351e3c25b03e5267130b9db7 adds it back in, and explains why. The following commit adds the equivalent packages for amd64 cross compilation from aarch64 Linux

I'm not sure if `musl-tools` is still necessary, but also added that to both help texts. I will test further, but overall it's probably best to keep the changes introduced by this pull request minimal (and improve on the help text in another PR).

Malax reviewed (changes requested):
Thanks for the changes! Looking great! 👍🏻 Let's add a brief `CHANGELOG` entry and ship it!

edmorley commented on CHANGELOG.md:
```diff
@@ -13,11 +13,15 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 - `libcnb`:
   - Changed `Layer` interface from `&self` to `&mut self`. ([#669](https://github.com/heroku/libcnb.rs/pull/669))
+- `libcnb-package`:
```
As of #662, we treat `libcnb-package` (and `libcnb-common`) as an internal packages that has no stability guarantees. As such, we don't mention them directly in the changelog, but instead mention any end-user impact in terms of what the user would experience - which in this case is via `libcnb-cargo` and `libcnb-test`.

However, rather than duplicate the entries under both a `libcnb-cargo` bullet and a `libcnb-test` bullet, what we did for the last cross-compile assistance change was to mention it at the top level, like so:
https://github.com/heroku/libcnb.rs/blob/2907c182b4aaf72d90a312e95455e3b51c0399a8/CHANGELOG.md?plain=1#L36

Could you remove the `libcnb-package` bullet and move the entries to the top level?

runesoerensen replied:
Ah didn't know we made that distinction (I modeled the entries after [this entry](https://github.com/heroku/libcnb.rs/pull/577/commits/fc4f442b5d47f30c85a92b9af32aac9ccab12f02), which seems to have been the last one prior to marking the `libcnb-package` internal :)).

I [moved the entries up](https://github.com/heroku/libcnb.rs/pull/769/commits/d71b6c5c53cdea63953379969b87a2872bc0af77) to the top level (both under the `Changed` header) without the `libcnb-package` bullet.

edmorley reviewed (approved):
Thank you for this! The implementation is both cleaner and more correct now :-)

I tested this manually locally using on macOS by packaging the Python CNB (which includes C deps, such as `ring` to exercise the env vars being set), in the following scenarios:
- the default target (`x86_64-unknown-linux-musl`) with and without the correct toolchain installed
- `--target aarch64-unknown-linux-musl` with and without the correct toolchain installed

...and it worked great for all of those.

(Longer term we have #730 and #731 for exploring trying to test more scenarios in CI)

> However, the binary used when the architectures don't match is still the *-gnu-gcc variant (as for instance aarch64-linux-musl-gcc is only available on aarch64 Linux (this requires further investigation, and a separate issue should be created for that)).

The intention of #724 was to be about the general case of "we probably shouldn't be using the non-MUSL gcc ever". This PR helps with a subset of cases, but not all (as you mentioned in the PR description). Rather than close #724 out, I've morphed the title to make it clearer that it's about the general case :-)

runesoerensen commented on libcnb-package/src/cross_compile.rs:
```diff
@@ -7,95 +9,89 @@ use which::which;
-        // toolchain.
-        let possible_gcc_binary_names = ["x86_64-unknown-linux-musl-gcc", "x86_64-linux-musl-gcc"];
+    let target_triple = target_triple.as_ref();
+    let (gcc_path, help_text) = match (target_triple, consts::OS, consts::ARCH) {
```
Agree - forgot to make that change, but was thinking the same thing! Addressed in https://github.com/heroku/libcnb.rs/pull/769/commits/e040fee6ff05cce89cc83312f64eb98dc3a5ce0a

edmorley commented on CHANGELOG.md:
```diff
@@ -11,6 +11,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
+- Improved the consistency of cross-compilation assistance provided across all supported `target_triple` and host OS/architecture combinations. [#769](https://github.com/heroku/libcnb.rs/pull/769)
+- Added cross-compilation assistance for `aarch64-unknown-linux-musl` (on macOS and ARM64 Linux) and `x86_64-unknown-linux-musl` (on ARM64 Linux). [#769](https://github.com/heroku/libcnb.rs/pull/769)
+
```
```suggestion
```

runesoerensen replied:
Lets pretend that never happened ;-) https://github.com/heroku/libcnb.rs/compare/d71b6c5c53cdea63953379969b87a2872bc0af77..f9fb1cfcc4030551941c860ad55004b44d90085b
